RepositoryCompiler/CodeModel/CodeParsers/CSharp/CSharpMetricCalculator.cs

```diff
@@ -20,10 +20,23 @@ public CaDETClassMetrics CalculateClassMetrics(CaDETClass parsedClass)
                 NAD = GetNumberOfAttributesDefined(parsedClass),
                 WMC = GetWeightedMethodPerClass(parsedClass),
                 LCOM = GetLackOfCohesionOfMethods(parsedClass),
-                TCC = GetTightClassCohesion(parsedClass)
+                TCC = GetTightClassCohesion(parsedClass),
+                ATFD = GetAccessToForeignData(parsedClass)
             };
         }
 
+        private int GetAccessToForeignData(CaDETClass parsedClass)
+        {
+            int numberOfAccessToForeignFieldsAndAccessors = 0;
+
+            foreach (var member in parsedClass.Members)
+            {
+                numberOfAccessToForeignFieldsAndAccessors += (member.AccessedFields.Count - member.GetAccessedOwnFields().Count) + (member.AccessedAccessors.Count - member.GetAccessedOwnAccessors().Count);
```

Comment:
I was of the opinion that we concluded that there were 3+ references that state that ATFD counts the number of external classes that we are accessing either through a field or accessor. Here we are counting the actual number of foreign fields and accessors accessed.

Is this correct? I remember that the thresholds for ATFD were between 2 and 5. Could you list the few papers (e.g., by name) based on which we are going with this calculation?

Reply:
@Pufke In the event that we are indeed tracking the different classes and not the fields, we should take inheritance in to account.
Namely, lets examine the following situation:
- Class A defines public field/accessor X, and class B inherits A and defines public field/accessor Y.
- Class C references B and accesses both Y and X.
- ATFD for C should be one.

We can write a test for this, in the event that ATFD will indeed count distinct classes.

Reply:
Yea, at the beginning of the research we had that opinion.

But after researching, we concluded the following: Access To Foreign Data (ATFD) is the number of
attributes in foreign classes accessed by class. 

 References from the following paper:
 Bad smell detection using quality metrics and refactoring
 opportunities 
 DOI: 10.1002/smr.2255

 References: 

-  Macia I, Garcia J, Popescu D, Garcia A, Medvidovic N, von Staa A. Are automatically-detected code 
   anomalies relevant to architectural modularity
   DOI: 10.1145/2162049.2162069
   Screenshot from paper: 
![image](https://user-images.githubusercontent.com/49925421/98228322-b89fde00-1f58-11eb-92f2-25802453d198.png)

-   The Evolution and Impact of Code Smells: A Case Study of Two Open Source Systems 
    DOI: 10.1109/ESEM.2009.5314231
    Screenshot from paper:
![image](https://user-images.githubusercontent.com/49925421/98228942-8e9aeb80-1f59-11eb-8d85-898219094a39.png)